<a href="https://colab.research.google.com/github/Divya-Dhananjayan/IPSR-internship/blob/main/Task2internship_DataCleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd


**Step 1: Data Cleaning**
  =====================

**Handling File 1**
  ---------------

In [ ]:
# Importing file containing details of Phd students

df1 = pd.read_csv('/content/nirf-admissions-ranking-and-academic-year-and-university-institute-wise-total-number-of-students-pursuing-part-and-full-time-ph-d-courses.csv')




In [ ]:
# To get an overview of the data just loaded the first few rows and header

df1.head()



ranking_year ranking_category institute_id  \
0          2024       University  IR-O-U-0250   
1          2024       University  IR-O-U-0250   
2          2024       University  IR-O-U-0249   
3          2024       University  IR-O-U-0249   
4          2024       University  IR-O-U-0043   

                                      institute_name           state  \
0                                Yenepoya University       Karnataka   
1                                Yenepoya University       Karnataka   
2              Visvesvaraya Technological University       Karnataka   
3              Visvesvaraya Technological University       Karnataka   
4  Vignan's Foundation for Science, Technology an...  Andhra Pradesh   

        city                                      academic_year   category  \
0  Mangaluru  Ph.D (Student pursuing doctoral program till 2...  Full Time   
1  Mangaluru  Ph.D (Student pursuing doctoral program till 2...  Part Time   
2    Belgaum  Ph.D (Student pursuing doctoral program till 2...  Full Time   
3    Belgaum  Ph.D (Student pursuing doctoral program till 2...  Part Time   
4     Guntur  Ph.D (Student pursuing doctoral program till 2...  Full Time   

   value                      unit  
0    122  value in Absolute Number  
1     42  value in Absolute Number  
2     57  value in Absolute Number  
3    224  value in Absolute Number  
4    502  value in Absolute Number

In [ ]:
#Checking data types of each columnn entries

df1.dtypes



ranking_year         int64
ranking_category    object
institute_id        object
institute_name      object
state               object
city                object
academic_year       object
category            object
value                int64
unit                object
dtype: object

Here academic year column is object type as it is having alphanumeric values we need to change it into date format after extracting the date alone.

In [ ]:
# Checking Statistical summary
df1.describe ( include="all")



ranking_year ranking_category institute_id  \
count    9342.000000             9342         9342   
unique           NaN               13         1320   
top              NaN      Engineering  IR-O-U-0497   
freq             NaN             2340           48   
mean     2021.253907              NaN          NaN   
std         1.922649              NaN          NaN   
min      2018.000000              NaN          NaN   
25%      2020.000000              NaN          NaN   
50%      2021.000000              NaN          NaN   
75%      2023.000000              NaN          NaN   
max      2024.000000              NaN          NaN   

                                    institute_name       state     city  \
count                                         9342        9342     9342   
unique                                         685          33      225   
top     Kalinga Institute of Industrial Technology  Tamil Nadu  Chennai   
freq                                            92        1474      632   
mean                                           NaN         NaN      NaN   
std                                            NaN         NaN      NaN   
min                                            NaN         NaN      NaN   
25%                                            NaN         NaN      NaN   
50%                                            NaN         NaN      NaN   
75%                                            NaN         NaN      NaN   
max                                            NaN         NaN      NaN   

                                            academic_year   category  \
count                                                9342       9342   
unique                                                  7          2   
top     Ph.D (Student pursuing doctoral program till 2...  Full Time   
freq                                                 1532       4671   
mean                                                  NaN        NaN   
std                                                   NaN        NaN   
min                                                   NaN        NaN   
25%                                                   NaN        NaN   
50%                                                   NaN        NaN   
75%                                                   NaN        NaN   
max                                                   NaN        NaN   

              value                      unit  
count   9342.000000                      9342  
unique          NaN                         1  
top             NaN  value in Absolute Number  
freq            NaN                      9342  
mean     251.983408                       NaN  
std      514.658732                       NaN  
min        0.000000                       NaN  
25%        5.000000                       NaN  
50%       50.000000                       NaN  
75%      255.000000                       NaN  
max     5432.000000                       NaN

In [ ]:
# Need a concise summary of the Dataframe

df1.info()




<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9342 entries, 0 to 9341
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ranking_year      9342 non-null   int64 
 1   ranking_category  9342 non-null   object
 2   institute_id      9342 non-null   object
 3   institute_name    9342 non-null   object
 4   state             9342 non-null   object
 5   city              9342 non-null   object
 6   academic_year     9342 non-null   object
 7   category          9342 non-null   object
 8   value             9342 non-null   int64 
 9   unit              9342 non-null   object
dtypes: int64(2), object(8)
memory usage: 730.0+ KB


Null values are not present

In [ ]:


# Checking Missing values columnwise

df1.isnull().sum()




ranking_year        0
ranking_category    0
institute_id        0
institute_name      0
state               0
city                0
academic_year       0
category            0
value               0
unit                0
dtype: int64

In [ ]:
#checking for dulpictes

df1.duplicated(keep= False)



0       False
1       False
2       False
3       False
4       False
        ...  
9337    False
9338    False
9339    False
9340    False
9341    False
Length: 9342, dtype: bool

In [ ]:
df1.drop_duplicates()
df1

ranking_year ranking_category        institute_id  \
0             2024       University         IR-O-U-0250   
1             2024       University         IR-O-U-0250   
2             2024       University         IR-O-U-0249   
3             2024       University         IR-O-U-0249   
4             2024       University         IR-O-U-0043   
...            ...              ...                 ...   
9337          2018     Architecture    IR-1-A-A-C-46330   
9338          2018     Architecture  IR-3-A-EMA-C-43708   
9339          2018     Architecture  IR-3-A-EMA-C-43708   
9340          2018     Architecture  IR-4-A-OEMA-U-0439   
9341          2018     Architecture  IR-4-A-OEMA-U-0439   

                                         institute_name           state  \
0                                   Yenepoya University       Karnataka   
1                                   Yenepoya University       Karnataka   
2                 Visvesvaraya Technological University       Karnataka   
3                 Visvesvaraya Technological University       Karnataka   
4     Vignan's Foundation for Science, Technology an...  Andhra Pradesh   
...                                                 ...             ...   
9337  Faculty of Architecture, Manipal Academy of Hi...       Karnataka   
9338                  College of Engineering Trivandrum          Kerala   
9339                  College of Engineering Trivandrum          Kerala   
9340                                    Anna University      Tamil Nadu   
9341                                    Anna University      Tamil Nadu   

           city                                      academic_year   category  \
0     Mangaluru  Ph.D (Student pursuing doctoral program till 2...  Full Time   
1     Mangaluru  Ph.D (Student pursuing doctoral program till 2...  Part Time   
2       Belgaum  Ph.D (Student pursuing doctoral program till 2...  Full Time   
3       Belgaum  Ph.D (Student pursuing doctoral program till 2...  Part Time   
4        Guntur  Ph.D (Student pursuing doctoral program till 2...  Full Time   
...         ...                                                ...        ...   
9337      Udupi  Ph.D (Student pursuing doctoral program till 2...  Part Time   
9338       Pune  Ph.D (Student pursuing doctoral program till 2...  Full Time   
9339       Pune  Ph.D (Student pursuing doctoral program till 2...  Part Time   
9340    Chennai  Ph.D (Student pursuing doctoral program till 2...  Full Time   
9341    Chennai  Ph.D (Student pursuing doctoral program till 2...  Part Time   

      value                      unit  
0       122  value in Absolute Number  
1        42  value in Absolute Number  
2        57  value in Absolute Number  
3       224  value in Absolute Number  
4       502  value in Absolute Number  
...     ...                       ...  
9337      1  value in Absolute Number  
9338      0  value in Absolute Number  
9339      0  value in Absolute Number  
9340      1  value in Absolute Number  
9341      5  value in Absolute Number  

[9342 rows x 10 columns]

There isn't any duplicate entries.

Now lets consider each column.
-------

1.By analysing the data type of each column all the columns are perfect except academic_year, here academic year  is object type because it is an alpha numeric column.We need to extract only the year



2.There are serious issues in the institute_name column.

while closely analysing the data there are spelling errors ,special characters used , sometimes some additional information like location also included in the same column , punctuation differences etc
These all errors one or the other are included  when the same institute name is entered multiple times under the same institute id.

This cause multiplr institute name for the same institute id . we need to resolve it


3.More or less the same issue is there in the state and city column . But in the present stage it is not important the location . if needed we can consider it later in the project

4.The value column is actually the no. of phd students. We must rename it accordingly

5.The column name unit is actually not necessary. we can drop it

 considering 'institute_name' column
 -------------------------


In [ ]:
# Extra spaces and special characters from the institute_name column is removed


df1['institute_name'] = df1['institute_name'].str.lower().str.replace(r'[^a-z0-9 ]', '', regex=True).str.strip()


# trying to make one to one institute_id and institute_name
# keeping the first occurance of institute name for the institute id

grpd = df1.groupby('institute_id')['institute_name'].first()


# replacing the institute name column with unique name of institute name

df1['institute_name']= df1['institute_id'].map(grpd)


#  Now  chances for duplicate entries are there


df1.drop_duplicates()

ranking_year ranking_category        institute_id  \
0             2024       University         IR-O-U-0250   
1             2024       University         IR-O-U-0250   
2             2024       University         IR-O-U-0249   
3             2024       University         IR-O-U-0249   
4             2024       University         IR-O-U-0043   
...            ...              ...                 ...   
9337          2018     Architecture    IR-1-A-A-C-46330   
9338          2018     Architecture  IR-3-A-EMA-C-43708   
9339          2018     Architecture  IR-3-A-EMA-C-43708   
9340          2018     Architecture  IR-4-A-OEMA-U-0439   
9341          2018     Architecture  IR-4-A-OEMA-U-0439   

                                         institute_name           state  \
0                                   yenepoya university       Karnataka   
1                                   yenepoya university       Karnataka   
2                 visvesvaraya technological university       Karnataka   
3                 visvesvaraya technological university       Karnataka   
4     vignans foundation for science technology and ...  Andhra Pradesh   
...                                                 ...             ...   
9337  faculty of architecture manipal academy of hig...       Karnataka   
9338                  college of engineering trivandrum          Kerala   
9339                  college of engineering trivandrum          Kerala   
9340                                    anna university      Tamil Nadu   
9341                                    anna university      Tamil Nadu   

           city                                      academic_year   category  \
0     Mangaluru  Ph.D (Student pursuing doctoral program till 2...  Full Time   
1     Mangaluru  Ph.D (Student pursuing doctoral program till 2...  Part Time   
2       Belgaum  Ph.D (Student pursuing doctoral program till 2...  Full Time   
3       Belgaum  Ph.D (Student pursuing doctoral program till 2...  Part Time   
4        Guntur  Ph.D (Student pursuing doctoral program till 2...  Full Time   
...         ...                                                ...        ...   
9337      Udupi  Ph.D (Student pursuing doctoral program till 2...  Part Time   
9338       Pune  Ph.D (Student pursuing doctoral program till 2...  Full Time   
9339       Pune  Ph.D (Student pursuing doctoral program till 2...  Part Time   
9340    Chennai  Ph.D (Student pursuing doctoral program till 2...  Full Time   
9341    Chennai  Ph.D (Student pursuing doctoral program till 2...  Part Time   

      value                      unit  
0       122  value in Absolute Number  
1        42  value in Absolute Number  
2        57  value in Absolute Number  
3       224  value in Absolute Number  
4       502  value in Absolute Number  
...     ...                       ...  
9337      1  value in Absolute Number  
9338      0  value in Absolute Number  
9339      0  value in Absolute Number  
9340      1  value in Absolute Number  
9341      5  value in Absolute Number  

[8916 rows x 10 columns]

See the number of entries decrease to 8916 from 9342

let's check the unique number of institute_id and institute_name in df1

In [ ]:
df1[['institute_id','institute_name']].nunique()

institute_id      1320
institute_name     662
dtype: int64

The difference in their number is beacuse many institutions are multidisciplinary institutes having same institute_id for all their departments

In [ ]:
#Extracting year from 'academic_year' column

df1['academic_year'] =df1['academic_year'].str.extract(r'(\d{4}[-_]\d{2})')
df1['academic_year']




0       2022-23
1       2022-23
2       2022-23
3       2022-23
4       2022-23
         ...   
9337    2016-17
9338    2016-17
9339    2016-17
9340    2016-17
9341    2016-17
Name: academic_year, Length: 9342, dtype: object

In [ ]:

# Renaming the 'value' column

df1.rename(columns= {'value':'no. of phd students'} ,inplace= True)



In [ ]:
# Removing the unit column

df1.drop('unit',axis=1,inplace = True)



In [ ]:
df1

ranking_year ranking_category        institute_id  \
0             2024       University         IR-O-U-0250   
1             2024       University         IR-O-U-0250   
2             2024       University         IR-O-U-0249   
3             2024       University         IR-O-U-0249   
4             2024       University         IR-O-U-0043   
...            ...              ...                 ...   
9337          2018     Architecture    IR-1-A-A-C-46330   
9338          2018     Architecture  IR-3-A-EMA-C-43708   
9339          2018     Architecture  IR-3-A-EMA-C-43708   
9340          2018     Architecture  IR-4-A-OEMA-U-0439   
9341          2018     Architecture  IR-4-A-OEMA-U-0439   

                                         institute_name           state  \
0                                   yenepoya university       Karnataka   
1                                   yenepoya university       Karnataka   
2                 visvesvaraya technological university       Karnataka   
3                 visvesvaraya technological university       Karnataka   
4     vignans foundation for science technology and ...  Andhra Pradesh   
...                                                 ...             ...   
9337  faculty of architecture manipal academy of hig...       Karnataka   
9338                  college of engineering trivandrum          Kerala   
9339                  college of engineering trivandrum          Kerala   
9340                                    anna university      Tamil Nadu   
9341                                    anna university      Tamil Nadu   

           city academic_year   category  no. of phd students  
0     Mangaluru       2022-23  Full Time                  122  
1     Mangaluru       2022-23  Part Time                   42  
2       Belgaum       2022-23  Full Time                   57  
3       Belgaum       2022-23  Part Time                  224  
4        Guntur       2022-23  Full Time                  502  
...         ...           ...        ...                  ...  
9337      Udupi       2016-17  Part Time                    1  
9338       Pune       2016-17  Full Time                    0  
9339       Pune       2016-17  Part Time                    0  
9340    Chennai       2016-17  Full Time                    1  
9341    Chennai       2016-17  Part Time                    5  

[9342 rows x 9 columns]

In [ ]:


# exporting cleaned df1 to my local system as NIRF_File_1

df1.to_csv("NIRF_File_1.csv",index= False)






***Handling File 2 ***
==============


In [ ]:

# Loading second file having score and rank of institutes

df2=pd.read_csv('/content/nirf-rankings-ranking-year-and-university-institute-wise-total-score-and-rank-of-educational-institutions.csv')



In [ ]:

# Loading first few rows
df2.head()



ranking_year ranking_category  institute_id  \
0          2024              Law   IR-L-U-0585   
1          2024           Dental   IR-N-U-0078   
2          2024           Dental   IR-N-U-0169   
3          2024           Dental  IR-N-C-40193   
4          2024           Dental  IR-N-C-35422   

                                      institute_name        city        state  \
0  The West Bengal National University of Juridic...     Kolkata  West Bengal   
1                                  Panjab University  Chandigarh   Chandigarh   
2  Manav Rachna International Institute of Resear...   Faridabad      Haryana   
3                   Bapuji Dental College & Hospital   Davangere    Karnataka   
4  Bharati Vidyapeeth (Deemed to be University) D...        Pune  Maharashtra   

   score  rank  
0  76.39   4.0  
1  52.99  27.0  
2  49.88  38.0  
3  50.86  37.0  
4  51.06  36.0

In [ ]:

# Checking data types
df2.dtypes


ranking_year          int64
ranking_category     object
institute_id         object
institute_name       object
city                 object
state                object
score               float64
rank                 object
dtype: object

Here the column rank is object type but it should be of int type. we need to change it for further analysis

In [ ]:

# Checking statistical summary
df2.describe(include="all")



ranking_year ranking_category institute_id             institute_name  \
count    6340.000000             6340         6340                       6340   
unique           NaN               18         2418                       1558   
top              NaN      Engineering  IR-O-U-0098  Aligarh Muslim University   
freq             NaN             1300           13                         57   
mean     2020.694953              NaN          NaN                        NaN   
std         2.384946              NaN          NaN                        NaN   
min      2016.000000              NaN          NaN                        NaN   
25%      2019.000000              NaN          NaN                        NaN   
50%      2021.000000              NaN          NaN                        NaN   
75%      2023.000000              NaN          NaN                        NaN   
max      2024.000000              NaN          NaN                        NaN   

           city       state        score  rank  
count      6340        6340  6313.000000  6340  
unique      453          33          NaN   308  
top     Chennai  Tamil Nadu          NaN   1.0  
freq        440        1188          NaN    85  
mean        NaN         NaN    51.971948   NaN  
std         NaN         NaN    11.032680   NaN  
min         NaN         NaN    30.310000   NaN  
25%         NaN         NaN    44.220000   NaN  
50%         NaN         NaN    51.070000   NaN  
75%         NaN         NaN    57.800000   NaN  
max         NaN         NaN    94.460000   NaN

Here the min ,max,quartile entries of rank column is not shown so it means some problems are there in that column. We have seen it as object type earlier.

In [ ]:

df2.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6340 entries, 0 to 6339
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ranking_year      6340 non-null   int64  
 1   ranking_category  6340 non-null   object 
 2   institute_id      6340 non-null   object 
 3   institute_name    6340 non-null   object 
 4   city              6340 non-null   object 
 5   state             6340 non-null   object 
 6   score             6313 non-null   float64
 7   rank              6340 non-null   object 
dtypes: float64(1), int64(1), object(6)
memory usage: 396.4+ KB


From this it is understood that missing values are there in the score column

In [ ]:

# calculating how many missing values are there

df2.isnull().sum()


ranking_year         0
ranking_category     0
institute_id         0
institute_name       0
city                 0
state                0
score               27
rank                 0
dtype: int64

The score column have 27 null values but the ranking column have all the values. These two columns are closely related in all terms. So we need to work on that for handling the missing values

Now consider each column
-----

1.There are serious issues in the institute_name column.

while closely analysing the data there are spelling errors ,special characters used , sometimes some additional information like location also included in the same column , punctuation differences etc These all errors one or the other are included when the same institute name is entered multiple times under the same institute id.

This cause multiplr institute name for the same institute id . we need to resolve it


2.More or less the same issue is there in the state and city column . But in the present stage it is not important the location . if needed we can consider it later in the project

3.There are Missing values in the 'score' column

4.The rank column is of object type . we need to change it to int

In [ ]:
# checking number of unique institute id in df2

df2['institute_id'].nunique()


2418

In the second file having score and rank of institutes have data from 2016 onwards. And the category field have additional 5 entries compared to first file . Here the data is more

we know there are complicated issues in the institute name column , so my plan is to replace the institute name of df2 with institute name of df1 according to the institute id.Both the files have institute id same .

In [ ]:
# Making a data frame having  the 'institute_id' and the 'institute_name' columns
#columns from df1

df1_institute_id_name = df1[['institute_id','institute_name']]

# AS we  are considering only two parameters id and name our new data frame
# have the same rows multiple times. So we must remove the duplicate entries

df1_institute_id_name = df1_institute_id_name.drop_duplicates(subset= 'institute_id')

# restoring the result to the same dataframe

df1_institute_id_name

institute_id                                     institute_name
0             IR-O-U-0250                                yenepoya university
2             IR-O-U-0249              visvesvaraya technological university
4             IR-O-U-0043  vignans foundation for science technology and ...
6             IR-O-U-0490                    vellore institute of technology
8             IR-O-U-0489  vel tech rangarajan dr sagunthala r and d inst...
...                   ...                                                ...
9332   IR-4-A-OEMA-U-0560             indian institute of technology roorkee
9334  IR-5-A-OEMAL-U-0573           indian institute of technology kharagpur
9336     IR-1-A-A-C-46330  faculty of architecture manipal academy of hig...
9338   IR-3-A-EMA-C-43708                  college of engineering trivandrum
9340   IR-4-A-OEMA-U-0439                                    anna university

[1320 rows x 2 columns]


Now we got it right. But we are trying to map this to df2. Map function works only on series. Not with dataframe so we convert it to series.



In [ ]:

# creating series having institute id as index and column as institute-name

id_correspondingto_name = df1_institute_id_name.set_index('institute_id')['institute_name']
id_correspondingto_name



institute_id
IR-O-U-0250                                          yenepoya university
IR-O-U-0249                        visvesvaraya technological university
IR-O-U-0043            vignans foundation for science technology and ...
IR-O-U-0490                              vellore institute of technology
IR-O-U-0489            vel tech rangarajan dr sagunthala r and d inst...
                                             ...                        
IR-4-A-OEMA-U-0560                indian institute of technology roorkee
IR-5-A-OEMAL-U-0573             indian institute of technology kharagpur
IR-1-A-A-C-46330       faculty of architecture manipal academy of hig...
IR-3-A-EMA-C-43708                     college of engineering trivandrum
IR-4-A-OEMA-U-0439                                       anna university
Name: institute_name, Length: 1320, dtype: object

In [ ]:


# Now replacing the institute name column of df2
#filling institute name considering the institute id and filling accordingly with
# the corresponding institute name from df1



df2['institute_name'] = df2['institute_id'].map(id_correspondingto_name).fillna(df2['institute_name'])



In [ ]:

# dropping duplicates

df2.drop_duplicates()


ranking_year ranking_category            institute_id  \
0             2024              Law             IR-L-U-0585   
1             2024           Dental             IR-N-U-0078   
2             2024           Dental             IR-N-U-0169   
3             2024           Dental            IR-N-C-40193   
4             2024           Dental            IR-N-C-35422   
...            ...              ...                     ...   
6335          2016       University           NIRF-UNIV-467   
6336          2016       University            NIRF-UNIV-15   
6337          2016       University           NIRF-UNIV-350   
6338          2016       University           NIRF-UNIV-540   
6339          2016         Pharmacy  NIRF-PHRM-1-2451478600   

                                         institute_name             city  \
0     the west bengal national university of juridic...          Kolkata   
1                                     punjab university       Chandigarh   
2     manav rachna international institute of resear...        Faridabad   
3                    bapuji dental college and hospital        Davangere   
4     bharati vidyapeeth deemed to be university den...             Pune   
...                                                 ...              ...   
6335  National Institute Of Food Technology, Entrepr...          Sonepat   
6336                                Alagappa University        Karaikudi   
6337  Karunya Institute Of Technology And Sciences-C...       Coimbatore   
6338                                 Periyar University            Salem   
6339  Periyar College Of Pharmaceuitcal Sciences-Tir...  TIRUCHIRAPPALLI   

            state  score  rank  
0     West Bengal  76.39   4.0  
1      Chandigarh  52.99  27.0  
2         Haryana  49.88  38.0  
3       Karnataka  50.86  37.0  
4     Maharashtra  51.06  36.0  
...           ...    ...   ...  
6335      Haryana  60.44    50  
6336   Tamil Nadu  60.63    49  
6337   Tamil Nadu  60.85    48  
6338   Tamil Nadu  60.99    47  
6339   Tamil Nadu  53.17  27.0  

[6340 rows x 8 columns]

Now we got a cleaned institute name column


Now consider the score column

In [ ]:

# handling missing values in score column

df2[df2['score'].isna()]


ranking_year  ranking_category institute_id  \
85            2024        Innovation  IR-I-U-0573   
96            2024        Innovation  IR-I-U-0439   
97            2024        Innovation  IR-I-U-0184   
99            2024        Innovation  IR-I-I-1074   
100           2024        Innovation  IR-I-U-0560   
101           2024        Innovation  IR-I-I-1075   
102           2024        Innovation  IR-I-U-0220   
103           2024        Innovation  IR-I-U-0013   
104           2024        Innovation  IR-I-U-0456   
105           2024        Innovation  IR-I-U-0306   
300           2024  Skill University  IR-S-U-0979   
301           2024  Skill University  IR-S-U-1048   
302           2024  Skill University  IR-S-U-0949   
428           2024   Open University  IR-V-U-0131   
429           2024   Open University  IR-V-U-0578   
430           2024   Open University  IR-V-U-0104   
1685          2023        Innovation   ARI-I-1075   
1686          2023        Innovation   ARI-U-0456   
1687          2023        Innovation   ARI-U-0013   
1688          2023        Innovation   ARI-I-1074   
1689          2023        Innovation   ARI-U-0560   
1690          2023        Innovation   ARI-U-0220   
1691          2023        Innovation   ARI-U-0306   
1692          2023        Innovation   ARI-U-0455   
1693          2023        Innovation   ARI-U-0263   
1694          2023        Innovation   ARI-U-0701   
1776          2023        Innovation   ARI-I-1075   

                                         institute_name       city  \
85             Indian Institute of Technology Kharagpur  Kharagpur   
96                                      Anna University    Chennai   
97                 Indian Institute of Technology Mandi      Mandi   
99                 Indian Institute of Technology Delhi  New Delhi   
100              Indian Institute of Technology Roorkee    Roorkee   
101               Indian Institute of Technology Kanpur     Kanpur   
102              Indian Institute of Science, Bengaluru  Bengaluru   
103            Indian Institute of Technology Hyderabad  Hyderabad   
104               Indian Institute of Technology Madras    Chennai   
105               Indian Institute of Technology Bombay     Mumbai   
300   Symbiosis Skills and Professional University (...       Pune   
301                   Shri Vishwakarma Skill University     Palwal   
302               Bhartiya Skill Development University  Rajasthan   
428         dr babasaheb ambedkar marathwada university  Ahmedabad   
429                      netaji subhash open university    Kolkata   
430        indira gandhi national open university ignou  New Delhi   
1685              INDIAN INSTITUTE OF TECHNOLOGY KANPUR     Kanpur   
1686              INDIAN INSTITUTE OF TECHNOLOGY MADRAS    Chennai   
1687           INDIAN INSTITUTE OF TECHNOLOGY HYDERABAD  Hyderabad   
1688               INDIAN INSTITUTE OF TECHNOLOGY DELHI  New Delhi   
1689             INDIAN INSTITUTE OF TECHNOLOGY ROORKEE    Roorkee   
1690             INDIAN INSTITUTE OF SCIENCE, BENGALURU  Bengaluru   
1691              INDIAN INSTITUTE OF TECHNOLOGY BOMBAY     Mumbai   
1692  INDIAN INSTITUTE OF INFORMATION TECHNOLOGY, DE...    Chennai   
1693           NATIONAL INSTITUTE OF TECHNOLOGY CALICUT  Kozhikode   
1694  INDIAN INSTITUTE OF TECHNOLOGY (BANARAS HINDU ...   Varanasi   
1776              INDIAN INSTITUTE OF TECHNOLOGY KANPUR   Varanasi   

                 state  score  rank  
85         West Bengal    NaN   9.0  
96          Tamil Nadu    NaN  10.0  
97    Himachal Pradesh    NaN   8.0  
99               Delhi    NaN   7.0  
100        Uttarakhand    NaN   6.0  
101      Uttar Pradesh    NaN   5.0  
102          Karnataka    NaN   4.0  
103          Telangana    NaN   3.0  
104         Tamil Nadu    NaN   2.0  
105        Maharashtra    NaN   1.0  
300        Maharashtra    NaN   1.0  
301            Haryana    NaN   2.0  
302          Rajasthan    NaN   3.0  
428            Gujarat    Na

while analysing the score column I identified that for the ranking category 'Innovation','Skill University','Open University' they dont have any score data even in the original data from the website. So we can consider it separately.

We can keep the missing data as missing data itself.

Considering the rank column

df2["rank"].astype(int)

 This line of code is used to change the column data type, but it is showing error. From this it is understood that it has some non integer values. we need to find out that.

Trying to figure out the nonint values in the rank column




In [ ]:
# Trying to convert 'rank' column to numeric . As it is not directly possible
# (we have seen before witth df2["rank"].astype(int) ) so convert  those values to nan

convert_to_numeric = pd.to_numeric(df2['rank'], errors = 'coerce')

# a boolean mask created

rows = convert_to_numeric.isna() | (convert_to_numeric % 1 !=0)

# rows will be true for non numeric and flaot valus

non_numeric_values = df2.loc[rows,'rank']

print(non_numeric_values)

5185               26A
5191               21A
5299               16A
5322    Rank Withdrawn
6063               20A
6164               31A
6216               16A
6219               14A
6307               41A
Name: rank, dtype: object


We need to work on this 9 values

The values 26A,21A and so on means there are multiple institues have same rank so keeping one as 21 and the coming ones as 21A, 21B like that the pattern is following. So we can ignore the Suffix

In [ ]:
# removing A for converting rank column to int type

df2['rank']=df2['rank'].str.replace(r'A$',"",regex = True)



In [ ]:
# Drop the row having Rank withdrawn
df2 = df2[df2['rank'] != 'Rank Withdrawn']

We must study about this institution how they lost their rank, It may help our project

In [ ]:
df2['rank'].dtypes

dtype('O')

In [ ]:
# we cannt convert a float type or non numeric object datatype directly to int
# here df2['rank'].astype(int) raising error

df2["rank"] = pd.to_numeric(df2["rank"], errors="coerce").astype("Int64")



In [ ]:
# Now checking the data type

df2['rank'].dtypes

Int64Dtype()

In [ ]:
# Saving the cleaned file to my local system

df2.to_csv('NIRF_File2.csv')

In [ ]:
df2

ranking_year ranking_category            institute_id  \
0             2024              Law             IR-L-U-0585   
1             2024           Dental             IR-N-U-0078   
2             2024           Dental             IR-N-U-0169   
3             2024           Dental            IR-N-C-40193   
4             2024           Dental            IR-N-C-35422   
...            ...              ...                     ...   
6335          2016       University           NIRF-UNIV-467   
6336          2016       University            NIRF-UNIV-15   
6337          2016       University           NIRF-UNIV-350   
6338          2016       University           NIRF-UNIV-540   
6339          2016         Pharmacy  NIRF-PHRM-1-2451478600   

                                         institute_name             city  \
0     the west bengal national university of juridic...          Kolkata   
1                                     punjab university       Chandigarh   
2     manav rachna international institute of resear...        Faridabad   
3                    bapuji dental college and hospital        Davangere   
4     bharati vidyapeeth deemed to be university den...             Pune   
...                                                 ...              ...   
6335  National Institute Of Food Technology, Entrepr...          Sonepat   
6336                                Alagappa University        Karaikudi   
6337  Karunya Institute Of Technology And Sciences-C...       Coimbatore   
6338                                 Periyar University            Salem   
6339  Periyar College Of Pharmaceuitcal Sciences-Tir...  TIRUCHIRAPPALLI   

            state  score  rank  
0     West Bengal  76.39     4  
1      Chandigarh  52.99    27  
2         Haryana  49.88    38  
3       Karnataka  50.86    37  
4     Maharashtra  51.06    36  
...           ...    ...   ...  
6335      Haryana  60.44    50  
6336   Tamil Nadu  60.63    49  
6337   Tamil Nadu  60.85    48  
6338   Tamil Nadu  60.99    47  
6339   Tamil Nadu  53.17    27  

[6339 rows x 8 columns]

Handling File 3
===========

In [ ]:
# importing file containing score category and value

df3 = pd.read_csv('/content/nirf-rankings-ranking-year-university-institute-and-category-wise-total-score-of-educational-institutions.csv')




In [ ]:
# Just loaded the first few entries

df3.head()

ranking_year ranking_category  institute_id  \
0          2024      Engineering   IR-E-U-0055   
1          2024          Medical     IR-D-N-15   
2          2024          Medical  IR-D-C-45654   
3          2024          Medical   IR-D-U-0236   
4          2024          Medical   IR-D-U-0368   

                                      institute_name        city       state  \
0           National Institute of Technology Silchar     Silchar       Assam   
1      All India Institute of Medical Sciences Delhi   New Delhi       Delhi   
2                          Christian Medical College     Vellore  Tamil Nadu   
3  National Institute of Mental Health and Neuro ...   Bengaluru   Karnataka   
4  Jawaharlal Institute of Post Graduate Medical ...  Puducherry  Puducherry   

             score_category  value  
0                Perception  10.06  
1  Outreach and Inclusivity  71.29  
2  Outreach and Inclusivity  62.69  
3  Outreach and Inclusivity  70.82  
4  Outreach and Inclusivity  67.81

In [ ]:
# loading summary statistics of all categories

df3.describe(include = "all")

ranking_year ranking_category institute_id             institute_name  \
count   70296.000000            70296        70296                      70296   
unique           NaN               15         2392                       1542   
top              NaN      Engineering  IR-O-U-0260  Aligarh Muslim University   
freq             NaN            15000          138                        630   
mean     2020.424889              NaN          NaN                        NaN   
std         2.279014              NaN          NaN                        NaN   
min      2016.000000              NaN          NaN                        NaN   
25%      2019.000000              NaN          NaN                        NaN   
50%      2021.000000              NaN          NaN                        NaN   
75%      2022.000000              NaN          NaN                        NaN   
max      2024.000000              NaN          NaN                        NaN   

           city       state score_category         value  
count     70296       70296          70296  70296.000000  
unique      451          33              9           NaN  
top     Chennai  Tamil Nadu     Perception           NaN  
freq       4854       13230          11716           NaN  
mean        NaN         NaN            NaN     50.312344  
std         NaN         NaN            NaN     22.644503  
min         NaN         NaN            NaN      0.000000  
25%         NaN         NaN            NaN     37.317500  
50%         NaN         NaN            NaN     53.870000  
75%         NaN         NaN            NaN     66.140000  
max         NaN         NaN            NaN    100.000000

In [ ]:
#checking data type of each column

df3.dtypes


ranking_year          int64
ranking_category     object
institute_id         object
institute_name       object
city                 object
state                object
score_category       object
value               float64
dtype: object

Data types are perfect

In [ ]:
# checking null values number

df3.isnull().sum()

ranking_year        0
ranking_category    0
institute_id        0
institute_name      0
city                0
state               0
score_category      0
value               0
dtype: int64

No null values

In [ ]:
df3.drop_duplicates()

ranking_year ranking_category            institute_id  \
0              2024      Engineering             IR-E-U-0055   
1              2024          Medical               IR-D-N-15   
2              2024          Medical            IR-D-C-45654   
3              2024          Medical             IR-D-U-0236   
4              2024          Medical             IR-D-U-0368   
...             ...              ...                     ...   
70109          2016      Engineering       NIRF-ENGG-INF-141   
70110          2016      Engineering  NIRF-ENGG-1-2451386658   
70111          2016      Engineering       NIRF-ENGG-INF-368   
70112          2016      Engineering  NIRF-ENGG-1-2449912182   
70113          2016      Engineering  NIRF-ENGG-1-2454054346   

                                          institute_name            city  \
0               National Institute of Technology Silchar         Silchar   
1          All India Institute of Medical Sciences Delhi       New Delhi   
2                              Christian Medical College         Vellore   
3      National Institute of Mental Health and Neuro ...       Bengaluru   
4      Jawaharlal Institute of Post Graduate Medical ...      Puducherry   
...                                                  ...             ...   
70109          National Institute Of Technology, Silchar         Silchar   
70110           Vishwakarma Institute Of Technology-Pune            PUNE   
70111           National Institute Of Technology, Raipur          Raipur   
70112  Bannari Amman Institute Of Technology-Sathyama...  SATHYAMANGALAM   
70113  Bharati Vidyapeeth Deemed University College O...            PUNE   

              state                    score_category  value  
0             Assam                        Perception  10.06  
1             Delhi          Outreach and Inclusivity  71.29  
2        Tamil Nadu          Outreach and Inclusivity  62.69  
3         Karnataka          Outreach and Inclusivity  70.82  
4        Puducherry          Outreach and Inclusivity  67.81  
...             ...                               ...    ...  
70109         Assam  Teaching, Learning and Resources  55.75  
70110   Maharashtra  Teaching, Learning and Resources  43.67  
70111  Chhattisgarh  Teaching, Learning and Resources  46.48  
70112    Tamil Nadu  Teaching, Learning and Resources  51.61  
70113   Maharashtra  Teaching, Learning and Resources  55.60  

[37878 rows x 8 columns]

Now consider each column
--------


1.1.There are serious issues in the institute_name column.

while closely analysing the data there are spelling errors ,special characters used , sometimes some additional information like location also included in the same column , punctuation differences etc These all errors one or the other are included when the same institute name is entered multiple times under the same institute id.

This cause multiplr institute name for the same institute id . we need to resolve it

2.More or less the same issue is there in the state and city column . But in the present stage it is not important the location . if needed we can consider it later in the project

3.Consider the score category column. If we get the 9 unique entries in the column as new 9 columns  with corresponding values as entries it would be better for analysing the data.  

4.Rename the value column as Category score it will be more understandable.

In [ ]:
# renaming value column

df3.rename(columns= {'value' : 'category_score'}, inplace= True)

In [ ]:
# Splitting the score_category  column to 9 new columns

df3_splitted = df3.pivot_table(
    index=['ranking_year','ranking_category','institute_id', 'institute_name'],
    columns='score_category',
    values='category_score'
).reset_index()

df3_splitted

score_category  ranking_year ranking_category            institute_id  \
0                       2016      Engineering  NIRF-ENGG-1-2449860060   
1                       2016      Engineering  NIRF-ENGG-1-2449883879   
2                       2016      Engineering  NIRF-ENGG-1-2449912182   
3                       2016      Engineering  NIRF-ENGG-1-2449920071   
4                       2016      Engineering  NIRF-ENGG-1-2449931180   
...                      ...              ...                     ...   
6308                    2024       University             IR-O-U-0575   
6309                    2024       University             IR-O-U-0642   
6310                    2024       University             IR-O-U-0739   
6311                    2024       University             IR-O-U-0747   
6312                    2024       University             IR-O-U-0749   

score_category                                     institute_name  \
0                    C.V.Raman College Of Engineering-Bhubaneswar   
1                                  Bengal Institute Of Technology   
2               Bannari Amman Institute Of Technology-Sathyama...   
3               Amrita Viswa Vidyapeetham-Amrita Nagar (Po) ,E...   
4                   Coimbatore Institute Of Technology-Coimbatore   
...                                                           ...   
6308                                          Jadavpur University   
6309                                        Shiv Nadar University   
6310                Madan Mohan Malaviya University of Technology   
6311                                        Chandigarh University   
6312                                   Manipal University, Jaipur   

score_category  Graduation Outcomes  Outreach and Inclusivity  Overall Score  \
0                             81.69                     55.36          53.91   
1                             92.06                     47.01          50.33   
2                             77.22                     62.78          56.21   
3                             81.38                     65.70          58.78   
4                             82.91                     68.54          62.58   
...                             ...                       ...            ...   
6308                          89.92                     49.94          65.39   
6309                          59.74                     66.79          49.80   
6310                          63.84                     70.12          46.33   
6311                          78.18                     77.52          59.96   
6312                          63.73                     66.15          49.50   

score_category  Perception  Qualitative Research  Quantitative Research  \
0                    23.00                   NaN                    NaN   
1                    59.00                   NaN                    NaN   
2                    27.00                   NaN                    NaN   
3                    36.00                   NaN                    NaN   
4                    74.00                   NaN                    NaN   
...                    ...                   ...                    ...   
6308                 63.74                   NaN                    NaN   
6309                 20.86                   NaN                    NaN   
6310                 19.63                   NaN                    NaN   
6311                 41.19                   NaN                    NaN   
6312                 27.16                   NaN                    NaN   

score_category  Research and Professional Practice  \
0                                            56.85   
1                                            50.00   
2                                            58.11   
3                                            53.42   
4                                            62.54   
...                                            ...   
6308                                         51.35   


Considering the institute_name column . i am trying to replace the institute name corresponding to the institute id with the institute name corresponding to the institute id of df1

In [ ]:
# replacing institute_name from df1
# just re using the series we prepared earlier from df1


df3_splitted['institute_name'] = df3_splitted['institute_id'].map(id_correspondingto_name).fillna(df3_splitted['institute_name'])

In [ ]:

#For cleaning purpose dropping the duplicates

df3_splitted.drop_duplicates()

score_category  ranking_year ranking_category            institute_id  \
0                       2016      Engineering  NIRF-ENGG-1-2449860060   
1                       2016      Engineering  NIRF-ENGG-1-2449883879   
2                       2016      Engineering  NIRF-ENGG-1-2449912182   
3                       2016      Engineering  NIRF-ENGG-1-2449920071   
4                       2016      Engineering  NIRF-ENGG-1-2449931180   
...                      ...              ...                     ...   
6308                    2024       University             IR-O-U-0575   
6309                    2024       University             IR-O-U-0642   
6310                    2024       University             IR-O-U-0739   
6311                    2024       University             IR-O-U-0747   
6312                    2024       University             IR-O-U-0749   

score_category                                     institute_name  \
0                    C.V.Raman College Of Engineering-Bhubaneswar   
1                                  Bengal Institute Of Technology   
2               Bannari Amman Institute Of Technology-Sathyama...   
3               Amrita Viswa Vidyapeetham-Amrita Nagar (Po) ,E...   
4                   Coimbatore Institute Of Technology-Coimbatore   
...                                                           ...   
6308                                          jadavpur university   
6309                                        shiv nadar university   
6310                madan mohan malaviya university of technology   
6311                                        chandigarh university   
6312                                    manipal university jaipur   

score_category  Graduation Outcomes  Outreach and Inclusivity  Overall Score  \
0                             81.69                     55.36          53.91   
1                             92.06                     47.01          50.33   
2                             77.22                     62.78          56.21   
3                             81.38                     65.70          58.78   
4                             82.91                     68.54          62.58   
...                             ...                       ...            ...   
6308                          89.92                     49.94          65.39   
6309                          59.74                     66.79          49.80   
6310                          63.84                     70.12          46.33   
6311                          78.18                     77.52          59.96   
6312                          63.73                     66.15          49.50   

score_category  Perception  Qualitative Research  Quantitative Research  \
0                    23.00                   NaN                    NaN   
1                    59.00                   NaN                    NaN   
2                    27.00                   NaN                    NaN   
3                    36.00                   NaN                    NaN   
4                    74.00                   NaN                    NaN   
...                    ...                   ...                    ...   
6308                 63.74                   NaN                    NaN   
6309                 20.86                   NaN                    NaN   
6310                 19.63                   NaN                    NaN   
6311                 41.19                   NaN                    NaN   
6312                 27.16                   NaN                    NaN   

score_category  Research and Professional Practice  \
0                                            56.85   
1                                            50.00   
2                                            58.11   
3                                            53.42   
4                                            62.54   
...                                            ...   
6308                                         51.35   


In [ ]:
# Saving the file to my local system

df3_splitted.to_csv('NIRF_File3.csv')



